test-harness.md

```diff
@@ -0,0 +1,94 @@
+MLS Interop Test Harness Specification
+======================================
+
+To automate interoperability testing across multiple MLS implementations, we
+need to be able to have a test runner control multiple clients, on different
+platforms and written in different languages. The interface to the MLS clients
+needs to cover all of the protocol operations for which we want to test interop,
+and expose enough information to evaluate whether interop succeeded.
+
+At a logical level, the necessary operations are all basically RPCs from the
+test runner to the clients.  First we describe the RPCs, then we discuss some
+ways these RPCs could be implemented.
+
+## Test Harness RPCs
+
+Ways to become a member of a group ("encrypt flag" indicates whether handshake
+messages in the group are encrypted):
+
+* Create a new group
+  * Inputs: Group ID, ciphersuite, credential, signature private key, encrypt
+    flag 
+  * Outputs: Pointer to new state
+* Create a new KeyPackage to begin a join
+  * Inputs: ciphersuite, credential, signature private key
+  * Outputs: KeyPackage, callback to which Welcome should be sent
+* Complete joining a group (Welcome callback)
+  * Inputs: Welcome, encrypt flag
+  * Outputs: Pointer to new state
+* External join
+  * Inputs: PublicGroupState, encrypt flag
+  * Outputs: MLSPlaintext(Commit), Welcome, pointer to new state
+
+Operations using a group state:
```

Comment:
Some of these operations (probably most of them) can also produce errors. In a second step it would be good to test implementations with faulty input data.

Reply:
I've marked everything as fallible, so we should be able to produce errors.  I think since we're passing in bytes here, we should also be able to pass in malformed bytes.  The main question is what malformed data we should provide and at what points, which I think is a good question for the "scenarios"